src/chess_engine.rs

```diff
@@ -1,19 +1,43 @@
 use rand::Rng;
-use shakmaty::{Move, MoveList, Position};
+use shakmaty::{Move, MoveList};
 use std::sync::mpsc::Sender;
 use std::sync::{Arc, RwLock};
 use std::thread;
 use std::time::Duration;
 
+use anyhow::Result as AnyResult;
+
 use log::{error, info};
 
 use crate::chess_game::ChessGame;
 use crate::websocket::Notification;
 
-pub trait ChooseMove {
-    fn choose_move(&self, fen: &str, legal_moves: &MoveList) -> Option<Move>;
+pub trait ChooseMove: Send + Sync {
+    /// Choose move has the assumption that the bot can play a move. Handing it a game with no legal moves
+    /// is invalid. The user should check if the game is over before calling this
+    fn choose_move(&self, chess_game: &ChessGame, legal_moves: &MoveList) -> AnyResult<Move>; // TODO: use something other than anyhow here plz
+}
+
+/// Allows us to distinguish between a saved bot and an active bot
+/// In addition Clone is not allowed for dyn trait objects and this helps us get around that
+pub trait ToChooseMove {
+    fn to_choose_move(&self) -> Box<dyn ChooseMove>;
 }
 
+impl<C: 'static + ChooseMove + Clone> ToChooseMove for C {
+    fn to_choose_move(&self) -> Box<dyn ChooseMove> {
+        Box::new(self.clone())
```

Comment:
Just to double check, this is for so that we can store anything which implements `ChooseMove` in our hashmap?

Reply:
Yeah so this is kinda messy but the idea is that we have a hashmap of bots that have been created we want to make a new instance of the bot so that if the bot is stateful there is a seperate instance for each active bot. So we just need to clone the `Box<dyn ChooseMove>` out of the struct but just adding `ChooseMove: Clone` as a trait bound makes it no longer object safe (not fully sure why that is. Still new to dynamic trait objects). This is kinda just a way to get around that limitation.

That all being said, none of the bots are stateful so we don't _strictly_ need this. We could just just `Arc<dyn ChooseMove>` in the hashmap and thene clone the arc. Then all active bots share the same underlying state but it does not matter right now.

Reply:
I see! The premise of a stateful bot is pretty interesting, are you envisioning something statistical like a Neural Network governing the moves made by the agent?

Reply:
I don't have anything in particular in mind. I'm just trying to allow for as much freedom in design as possible.

For example one could memosize heuristics for positions between calls to chooseMove